mondrian/src/main/java/mondrian/spi/impl/VerticaDialect.java

```diff
@@ -27,108 +31,165 @@
  */
 public class VerticaDialect extends JdbcDialectImpl {
 
-    public static final JdbcDialectFactory FACTORY =
-        new JdbcDialectFactory(
-            VerticaDialect.class,
-            DatabaseProduct.VERTICA);
-
-    /**
-     * Creates a VerticaDialect.
-     *
-     * @param connection Connection
-     */
-    public VerticaDialect(Connection connection) throws SQLException {
-        super(connection);
-    }
+  private final String flagsRegexp = "^(\\(\\?([a-zA-Z]+)\\)).*$";
```

Comment:
It looks like this regex for finding the flags gets duplicated in other dialects.  Could the flag logic be extracted and moved either into JdbcDialectImpl or some utility class?

Reply:
Thanks for reviewing Matt.  

It sure can.  I'm going to be working on getting regexp support for Redshift next so now is as good as time as ever to re-factor this a bit.

Reply:
Hey @bennychow I had this around my local box. I did some tests for Redshift and regular expressions. Hope this helps.

```
    @Override
    public boolean allowsRegularExpressionInWhereClause() {
        return true;
    }

    @Override
    public String generateRegularExpression(
        String source,
        String javaRegex)
    {
    	try {
            Pattern.compile(javaRegex);
        } catch (PatternSyntaxException e) {
            // Not a valid Java regex. Too risky to continue.
            return null;
        }

        // We might have to use case-insensitive matching
        javaRegex = DialectUtil.cleanUnicodeAwareCaseFlag(javaRegex);
        final Matcher flagsMatcher = flagsPattern.matcher(javaRegex);
        boolean caseSensitive = true;
        if (flagsMatcher.matches()) {
            final String flags = flagsMatcher.group(2);
            if (flags.contains("i")) {
                caseSensitive = false;
            }
        }
        if (flagsMatcher.matches()) {
            javaRegex =
                javaRegex.substring(0, flagsMatcher.start(1))
                + javaRegex.substring(flagsMatcher.end(1));
        }
        final Matcher escapeMatcher = escapePattern.matcher(javaRegex);
        while (escapeMatcher.find()) {
            javaRegex =
                javaRegex.replace(
                    escapeMatcher.group(1),
                    escapeMatcher.group(2));
        }
        final StringBuilder sb = new StringBuilder();

        // Now build the string.
        sb.append(source);
        sb.append(" IS NOT NULL AND ");
        
        // https://docs.aws.amazon.com/redshift/latest/dg/REGEXP_INSTR.html
        sb.append("REGEXP_INSTR(");
        sb.append(source);
        sb.append(",");
        quoteStringLiteral(sb, javaRegex);
        sb.append(",1,1,");
        sb.append(caseSensitive ? "c" : "i");
        sb.append(") > 0");

        return sb.toString();
    }
```

Reply:
Thanks Luc!  This will definitely be helpful.